[ConstraintSystem] Solve `where` clauses of for-in loops separately

xedin opened the pull request:
Doing so fits better into conjunction model which leads to more granular control
over what variables are brought into scope during `where` clause expression checking.

These changes also remove "one-way bind" flag from "for-in" statement target.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Solve `where` clauses of for-in loops separately

Doing so fits better into conjunction model which leads to more
granular control over what variables are brought into scope during
`where` clause expression checking.

These changes also remove "one-way bind" flag from "for-in" statement
target.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

hamishknight commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -959,6 +959,20 @@ class SyntacticElementConstraintGenerator
     // they would be handled together with pattern because pattern can
     // inform a type of sequence element e.g. `for i: Int8 in 0 ..< 8`
```
Looks like this comment needs tweaking slightly to not mention where clauses now

xedin replied:
Thanks!

xedin replied:
Done!

hamishknight commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -631,7 +631,7 @@ class SyntacticElementConstraintGenerator
     auto target = SyntacticElementTarget::forForEachStmt(
         forEachStmt, context.getAsDeclContext(),
-        /*bindTypeVarsOneWay=*/false);
+        /*ignoreWhereClause=*/true);
```
Would it be worth adding a comment here to explain that `where` clauses have already been handled by `visitForEachStmt`?

xedin replied:
Sure, I can add an explanation that points out that `visitForEachStmt` schedules a `where` clause as a separate element.

xedin replied:
Done!

hamishknight reviewed (approved):
Great to see more uses of `bindTypeVarsOneWay` being deleted 🎉

xedin committed:
[CSSyntacticElement] NFC: Update comments related to for-in where clause checking

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin committed:
[CSGen] Use `OptionalObject` constraint to connect element type and result of `next()`

Previously this wasn't possible because of `one-way` bind constraints
involved in pattern inference but this no longer the case.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility release
